File: proto/src/Rue/Proto/Prose.hs
{-# LANGUAGE OverloadedStrings #-}
-- | The verdict prose, docs/ROADMAP.md Appendix A: one clause per field
-- group, in this order: intent, reversibility, hold, point of no return,
-- gate, step gates, backstop, conditionals, controller-only undos, hosts
-- touched, dispatch, may-conflicts, unresolved bindings. A refused verdict
-- leads with its diagnostics instead.
module Rue.Proto.Prose
  ( prose
  ) where

import Data.Maybe (catMaybes, mapMaybe)
import Data.Text (Text)
import qualified Data.Text as T
import Rue.Proto.Diagnostics (codeText)
import Rue.Proto.Intent (Intent (..))
import Data.Aeson (Value (..))
import Rue.Proto.Model (renderDuration)
import Rue.Proto.Verdict

prose :: Verdict -> Text
prose v = case vStatus v of
  RefusedStatus ->
    vPlan v <> " on " <> vHost v <> ": refused; "
      <> T.intercalate "; " (map diag (vDiagnostics v))
      <> ".\n"
  Ok ->
    vPlan v <> " on " <> vHost v <> ": "
      <> T.intercalate "; " (concat clauses)
      <> ".\n"
  where
    diag d = codeText (diagCode d) <> maybe "" (\n -> " at step " <> num n) (diagStep d) <> ": " <> diagMessage d
    clauses =
      [ intentClause
      , [reversibility]
      , map holdClause (vHoldsAt v)
      , maybe [] (pure . knell) (vPointOfNoReturn v)
      , maybe [] (pure . gateClause) (vGate v)
      , mapMaybe stepGate (vSteps v)
      , maybe [] (pure . backstopClause) (vBackstop v)
      , conditionals
      , controllerOnly
      , hosts
      , deferred
      , [dispatch]
      , map mayConflict (vMayConflicts v)
      , unresolved
      ]
    num = T.pack . show
    intentClause = case vIntent v of
      Just Temporary -> ["temporary; reverts at wane " <> maybe "?" renderDuration (vWane v) <> induced]
      Just Permanent ->
        [ "permanent; commits at step " <> maybe "?" num (vCommitStep v)
            <> heldIndef
            <> (if vFiresByConstruction v then ", undo fires by construction" else "")
            <> induced
        ]
      Nothing -> ["intent undetermined"]
    heldIndef = case vHeldIndefinitely v of
      [] -> ""
      ns -> ", held indefinitely at step " <> steps ns <> " until an operator acts"
    induced = case vInducedDefer v of
      [] -> ""
      ns -> ", revert can be induced to defer at step " <> steps ns
    steps = T.intercalate ", " . map num
    total = length (vSteps v)
    reversibility
      | vReversibleThrough v == 0 = "not reversible past step 0"
      | vReversibleThrough v >= total && vPointOfNoReturn v == Nothing = "fully reversible (" <> num total <> " steps)"
      | otherwise = "reversible through step " <> num (vReversibleThrough v)
    -- Phase 0 finding: Appendix A has one hold clause, "(human required)";
    -- section 8.2 says a hold under mode: :auto holds "until resume, recant
    -- or commit", with no human in the loop, and the two cannot both be true.
    holdClause n
      | vMode v == "auto" = "step " <> num n <> " holds on refusal (until resume, recant or commit)"
      | otherwise = "step " <> num n <> " holds on refusal (human required)"
    knell p =
      "step " <> num (ponrStep p) <> " is a point of no return"
        <> maybe "" (", guard " <>) (ponrGuard p)
        <> ", cost " <> ponrCost p
        <> ", acknowledged by " <> ponrAck p
        <> maybe "" (\(f, t) -> "; step " <> num f <> " reversible back to step " <> num t) (vReversibleBackTo v)
    gateClause g
      | not (gvSatisfiable g) = "gate unsatisfiable"
      | gvZeroHumanPath g = "gate satisfiable with no human (allowed)"
      | otherwise = "gate satisfiable; minimum " <> maybe "?" num (gvMinDistinctHumans g) <> " distinct humans"
    stepGate s = case svGate s of
      Nothing -> Nothing
      Just g ->
        Just
          ( "step " <> num (svN s) <> " gated by " <> sgExpr g
              <> maybe "" (\d -> ", satisfiable by wait alone at +" <> renderDuration d) (sgWaitAloneAt g)
          )
    backstopClause b =
      "expiry backstop (" <> T.intercalate ", " (map triggerText (bvTriggers b)) <> ") covers "
        <> range (bvCovers b) <> " on the target"
        <> maybe "" (\n -> ", installed before step " <> num n) (bvInstalledBefore b)
        <> maybe "" (\n -> ", armed before step " <> num n) (bvArmedBefore b)
        <> maybe "" (\n -> ", armed after step " <> num n) (bvArmedAfter b)
        <> (case bvLateArmingWindow b of
              [] -> ""
              w -> ", engine-only for " <> range w <> " until armed")
        <> ", fires within ~" <> renderDuration (bvGranularity b) <> " after the deadline"
        <> (if bvSelfEnforced b then ", self-enforced on " <> vHost v else "")
        <> (case bvDriftPolicy b of
              [] -> ""
              ps -> ", drift: " <> T.intercalate ", " [num n <> " " <> p | (n, p) <- ps])
        <> ", snapshots on target (cap " <> num (bvSnapshotCap b) <> ")"
    triggerText t = case t of
      String s -> s
      other -> T.pack (show other)
    range ns = case ns of
      [] -> "no steps"
      [n] -> "step " <> num n
      _ -> "steps " <> num (minimum ns) <> "\x2013" <> num (maximum ns)
    conditionals =
      [ "step " <> num n <> " reverts unaided unless " <> c <> "; then deferred"
      | Just s <- [vBackstop v]
      , (n, c) <- bvConditional s
      ]
    controllerOnly = case vControllerOnlyUndos v of
      [] -> []
      [n] -> ["step " <> num n <> " reverts only while the engine lives"]
      ns -> ["steps " <> steps ns <> " revert only while the engine lives"]
    hosts =
      catMaybes
        [ case hs of
            [HostTouched h d] | h == vHost v && d == "target" -> Nothing
            [HostTouched "controller" _] -> Nothing
            _ -> Just ("step " <> num n <> " touches " <> T.intercalate ", " (map hostText hs))
        | (n, hs) <- vHostsTouched v
        ]
    hostText h = case h of
      HostTouched hn "controller" -> hn <> " (no instance directory; markers on controller)"
      HostTouched hn _ -> hn
      HostUnresolved _ -> "a host bound at runtime"
    deferred = case vDeferred v of
      [] -> []
      ns -> ["step " <> steps ns <> " deferred (handoff printed)"]
    dispatch = "clause dispatch from " <> dispatchSource (vDispatch v)
    mayConflict m =
      "may-conflict between steps " <> num (mcEarlier m) <> " and " <> num (mcLater m) <> " on " <> mcFact m
        <> (if mcRefused m then " (strict: refused)" else " (warn)")
    unresolved = case vUnresolvedBindings v of
      [] -> []
      bs -> ["unresolved binding(s): " <> T.intercalate ", " bs]

File: core/src/prose.rs
//! The verdict prose, docs/ROADMAP.md Appendix A: one clause per field group,
//! in this order: intent, reversibility, hold, point of no return, gate, step
//! gates, backstop, conditionals, controller-only undos, hosts touched,
//! deferred, dispatch, may-conflicts, unresolved bindings. A refused verdict
//! leads with its diagnostics instead.

use crate::intent::Intent;
use crate::verdict::*;

pub fn prose(v: &Verdict) -> String {
    let head = format!("{} on {}: ", v.plan, v.host);
    match v.status {
        Status::Refused => {
            let diags: Vec<String> = v
                .diagnostics
                .iter()
                .map(|d| {
                    format!(
                        "{}{}: {}",
                        d.code,
                        d.step.map(|n| format!(" at step {n}")).unwrap_or_default(),
                        d.message
                    )
                })
                .collect();
            format!("{head}refused; {}.\n", diags.join("; "))
        }
        Status::Ok => format!("{head}{}.\n", clauses(v).join("; ")),
    }
}

fn steps(ns: &[u32]) -> String {
    ns.iter()
        .map(|n| n.to_string())
        .collect::<Vec<_>>()
        .join(", ")
}

fn range(ns: &[u32]) -> String {
    match ns {
        [] => "no steps".to_string(),
        [n] => format!("step {n}"),
        _ => format!(
            "steps {}\u{2013}{}",
            ns.iter().min().unwrap(),
            ns.iter().max().unwrap()
        ),
    }
}

fn clauses(v: &Verdict) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();

    let induced = match v.induced_defer.as_slice() {
        [] => String::new(),
        ns => format!(", revert can be induced to defer at step {}", steps(ns)),
    };
    out.push(match v.intent {
        Some(Intent::Temporary) => format!(
            "temporary; reverts at wane {}{induced}",
            v.wane
                .map(|d| d.render())
                .unwrap_or_else(|| "?".to_string())
        ),
        Some(Intent::Permanent) => {
            let held = match v.held_indefinitely.as_slice() {
                [] => String::new(),
                ns => format!(
                    ", held indefinitely at step {} until an operator acts",
                    steps(ns)
                ),
            };
            format!(
                "permanent; commits at step {}{held}{}{induced}",
                v.commit_step
                    .map(|n| n.to_string())
                    .unwrap_or_else(|| "?".to_string()),
                if v.fires_by_construction {
                    ", undo fires by construction"
                } else {
                    ""
                }
            )
        }
        None => "intent undetermined".to_string(),
    });

    let total = v.steps.len() as u32;
    out.push(if v.reversible_through == 0 {
        "not reversible past step 0".to_string()
    } else if v.reversible_through >= total && v.point_of_no_return.is_none() {
        format!("fully reversible ({total} steps)")
    } else {
        format!("reversible through step {}", v.reversible_through)
    });

    // Phase 0 finding: Appendix A has one hold clause, "(human required)";
    // section 8.2 says a hold under mode: :auto holds "until resume, recant
    // or commit", with no human in the loop, and the two cannot both be true.
    for n in &v.holds_at {
        out.push(if v.mode == "auto" {
            format!("step {n} holds on refusal (until resume, recant or commit)")
        } else {
            format!("step {n} holds on refusal (human required)")
        });
    }

    if let Some(p) = &v.point_of_no_return {
        out.push(format!(
            "step {} is a point of no return{}, cost {}, acknowledged by {}{}",
            p.step,
            p.guard
                .as_ref()
                .map(|g| format!(", guard {g}"))
                .unwrap_or_default(),
            p.cost,
            p.ack,
            v.reversible_back_to
                .map(|(f, t)| format!("; step {f} reversible back to step {t}"))
                .unwrap_or_default()
        ));
    }

    if let Some(g) = &v.gate {
        out.push(if !g.satisfiable {
            "gate unsatisfiable".to_string()
        } else if g.zero_human_path {
            "gate satisfiable with no human (allowed)".to_string()
        } else {
            format!(
                "gate satisfiable; minimum {} distinct humans",
                g.min_distinct_humans
                    .map(|n| n.to_string())
                    .unwrap_or_else(|| "?".to_string())
            )
        });
    }

    for s in &v.steps {
        if let Some(g) = &s.gate {
            out.push(format!(
                "step {} gated by {}{}",
                s.n,
                g.expr,
                g.wait_alone_at
                    .map(|d| format!(", satisfiable by wait alone at +{}", d.render()))
                    .unwrap_or_default()
            ));
        }
    }

    if let Some(b) = &v.backstop {
        let mut c = format!(
            "expiry backstop ({}) covers {} on the target",
            b.triggers.join(", "),
            range(&b.covers)
        );
        if let Some(n) = b.installed_before {
            c.push_str(&format!(", installed before step {n}"));
        }
        if let Some(n) = b.armed_before {
            c.push_str(&format!(", armed before step {n}"));
        }
        if let Some(n) = b.armed_after {
            c.push_str(&format!(", armed after step {n}"));
        }
        if !b.late_arming_window.is_empty() {
            c.push_str(&format!(
                ", engine-only for {} until armed",
                range(&b.late_arming_window)
            ));
        }
        c.push_str(&format!(
            ", fires within ~{} after the deadline",
            b.granularity.render()
        ));
        if b.self_enforced {
            c.push_str(&format!(", self-enforced on {}", v.host));
        }
        if !b.drift_policy.is_empty() {
            c.push_str(&format!(
                ", drift: {}",
                b.drift_policy
                    .iter()
                    .map(|(n, p)| format!("{n} {p}"))
                    .collect::<Vec<_>>()
                    .join(", ")
            ));
        }
        c.push_str(&format!(", snapshots on target (cap {})", b.snapshot_cap));
        out.push(c);
        for (n, cond) in &b.conditional {
            out.push(format!(
                "step {n} reverts unaided unless {cond}; then deferred"
            ));
        }
    }

    match v.controller_only_undos.as_slice() {
        [] => {}
        [n] => out.push(format!("step {n} reverts only while the engine lives")),
        ns => out.push(format!(
            "steps {} revert only while the engine lives",
            steps(ns)
        )),
    }

    for (n, hs) in &v.hosts_touched {
        let skip = match hs.as_slice() {
            [HostTouched::Host { host, directory }] => {
                (host == &v.host && directory == "target") || host == "controller"
            }
            _ => false,
        };
        if skip {
            continue;
        }
        let texts: Vec<String> = hs
            .iter()
            .map(|h| match h {
                HostTouched::Host { host, directory } if directory == "controller" => {
                    format!("{host} (no instance directory; markers on controller)")
                }
                HostTouched::Host { host, .. } => host.clone(),
                HostTouched::Unresolved(_) => "a host bound at runtime".to_string(),
            })
            .collect();
        out.push(format!("step {n} touches {}", texts.join(", ")));
    }

    if !v.deferred.is_empty() {
        out.push(format!(
            "step {} deferred (handoff printed)",
            steps(&v.deferred)
        ));
    }

    out.push(format!("clause dispatch from {}", v.dispatch.source));

    for m in &v.may_conflicts {
        out.push(format!(
            "may-conflict between steps {} and {} on {}{}",
            m.earlier,
            m.later,
            m.fact,
            if m.refused {
                " (strict: refused)"
            } else {
                " (warn)"
            }
        ));
    }

    if !v.unresolved_bindings.is_empty() {
        out.push(format!(
            "unresolved binding(s): {}",
            v.unresolved_bindings.join(", ")
        ));
    }
    out
}
